In [3]:
#  scipy.stats.ttest_rel(Sample_1, Sample_2)
import numpy as np
import scipy
from scipy import stats
import scipy

# Session 5

## T Test
- For tests n < 30 are considered to be small sample tests
- Under the null hypothesis the test statistic follow exact distribution - t distribution, F distribution or χ2 distribution
- Small sample tests are regarded as exact tests
- for n == infinity it becomes the z test with the curve being normal 
- The T test is carried out under the assumption that the samples are drawn from independent normal population
- If the populations from which the samples are drawn do not have equal variance then the t-test can not be use. In this case the Welch test is used

Question: from ppt
</br>
A researcher is studying the growth of bacteria in waters of Lake Beach. The
mean bacteria count of 100 per unit volume of water is within the safety level.
The researcher collected 10 water samples of unit volume and found the mean
bacteria count to be 94.8 with a sample variance of 72.66. Does the data
indicate that the bacteria count is within the safety level? Test at the α = .05
level. Assume that the measurements constitute a sample from a normal
population.
</br>
Answer: </br>
- H0: u>100
- H1: u<100
- alpha = 0.05

- ------------------------
- conclusion:
 - since pvalue < alpha we reject the null hypthosis
 - t < t(9,0.05) => reject the null hypthosis

In [17]:
n = 10
x = 94.8
s = 72.66**0.5
t = (x-100)/(s/(n**0.5))
t_9_05 = scipy.stats.t.isf(0.05, df=9)
print(t, t_9_05)

pvalue = scipy.stats.t.cdf(t, df=9)
print(pvalue)

-1.9291040236750068 1.8331129326536337
0.04289782134327503


Question</br>
An experiment was conducted to compare the pain relieving hours of two new medicines.
Two groups of 14 and 15 patients were selected and were given comparable doses. Group
1 was given medicine 1 and group 2 was given medicine 2. Following data is obtained
from the two samples. Test whether the two populations give the same mean hours of
relief. Assume the data comes from normal distribution has equal variance. [Use α = 0.01]
</br>




In [18]:
x = 6.4
y = 7.3
s1 = 1.4
s2 = 1.5
n1 = 14
n2 = 15
s = (((n1-1)*(s1**2) + (n2-1)*(s2**2))/(n1+n2-2))**0.5
t = (x-y)/(s*(((1/n1)+(1/n2)))**0.5)

t_27_005 = scipy.stats.t.isf(0.005, df=27)

pvalue = 2 * scipy.stats.t.cdf(t, df=27)
print(t, t_27_005, pvalue)

print('pvalue > alpha => fail to reject null hypothesis')

-1.667146701465023 2.7706829571216756 0.10704510910964088
pvalue > alpha => fail to reject null hypothesis


An energy drink distributor claims that a new advertisement poster, featuring a life-size
picture of a well-known athlete. For a random sample of 10 outlets, the following data was
collected. Test that the null hypothesis that there the advertisement was effective in
increasing the sales




In [19]:
a = [33, 32, 38, 45, 37, 47, 48, 41, 45]
b = [42, 35, 31, 41, 37, 36, 49, 49, 48]

# diff = np.array(a) - np.array(b)
diff = np.array(b) - np.array(a)

d = diff.mean()
# std = diff.std()


# numpy guys are taking dinominator as n and not n-1 for std
s = np.std(diff)
s = s**2
s = (s*len(diff))/(len(diff)-1)
s = s**0.5

s = np.std(diff, ddof=1) # ddof = delta degrees of freedom

print(diff, d, s)

t = d/(s/len(diff)**0.5)
t

[  9   3  -7  -4   0 -11   1   8   3] 0.2222222222222222 6.610177338350647


np.float64(0.10085458113185983)

In [ ]:
# type 3 case
# alternative = 'two-sided' | 'less' | 'greater'
# arternative hypthosis is asserting that mean of a is greater than mean of b
scipy.stats.ttest_rel(a, b, alternative="greater")
scipy.stats.ttest_ind(a, b, alternative="greater") # better and more generic


TtestResult(statistic=np.float64(-0.0744838922723856), pvalue=np.float64(0.5292256731175016), df=np.float64(16.0))

In [ ]:

# accept or reject point
stats.t.isf(0.95, df=len(diff)-1)

np.float64(-1.8595480375228421)

In [ ]:
# pValue is probality that the null hypothesis is true
pValue = stats.t.cdf(t, df=len(diff)-1)
pValue

np.float64(0.5389261426537351)

In [36]:
??scipy.stats.ttest_rel


Signature:
scipy.stats.ttest_rel(
    a,
    b,
    axis=0,
    nan_policy='propagate',
    alternative='two-sided',
    *,
    keepdims=False,
)
Docstring:
Calculate the t-test on TWO RELATED samples of scores, a and b.

This is a test for the null hypothesis that two related or
repeated samples have identical average (expected) values.

Parameters
----------
a, b : array_like
    The arrays must have the same shape.
axis : int or None, default: 0
    If an int, the axis of the input along which to compute the statistic.
    The statistic of each axis-slice (e.g. row) of the input will appear in a
    corresponding element of the output.
    If ``None``, the input will be raveled before computing the statistic.
nan_policy : {'propagate', 'omit', 'raise'}
    Defines how to handle input NaNs.
    
    - ``propagate``: if a NaN is present in the axis slice (e.g. row) along
      which the  statistic is computed, the corresponding entry of the output
      will be NaN.
    - ``omit``: Na

From a sample 361 business owners had gone into bankruptcy due to recession. On taking
a survey, it was found that 105 of them had not consulted any professional for managing
their finance before opening the business. Test the null hypothesis that at most 25% of all
businesses had not consulted before opening the business.


modified problem...
H0 : P = 0.25
H1 : P != 0.25

bernoulli distribution used



In [ ]:
((105/361) - 0.25) / ((0.25*0.75/361)**0.5) # correct T value

1.7928245201151534

In [6]:
stats.norm.cdf(-1.7928245201151534) * 2

np.float64(0.07300098746249899)

From a sample 361 business owners had gone into bankruptcy due to recession. On taking
a survey, it was found that 105 of them had not consulted any professional for managing
their finance before opening the business. Test the null hypothesis that at most 25% of all
businesses had not consulted before opening the business.


H0 : P <= 0.25
H1 : P > 0.25



CDF (Cumulative Distribution Function): Probability of everything to the left of a point.
SF (Survival Function): Probability of everything to the right of a point.


In [7]:
stats.norm.sf(1.7928245201151534)

np.float64(0.03650049373124949)

### to clarify

Question
</br>
Steve owns a kiosk where sells two magazines - A and B in a month. He buys 100 copies of
magazine A out of which 78 were sold and 70 copies of magazine B out of which 65 were
sold. Is there enough evidence to say that magazine is B is more popular?
Test the claim using p-value technique. [Use α = 0.05].
</br>
Answer </br>


In [21]:
na = 100
nb = 70
xa = 78
xb = 65

pa = xa/na
pb = xb/nb

p = (xa + xb) / (na + nb)

z = (pa - pb) / ((p*(1-p))*((1/na)+(1/nb)))**0.5
z


-2.60830803458311

# ANOVA

In [ ]:
# one way ANOVA question slide ses 6
intra = 296.06
inter = 1778.0655

ts = (inter/3)/(intra/16)
ts

32.03072350199284

In [12]:
stats.f.isf(0.05, dfn=3, dfd=16)

np.float64(3.238871517453585)

In [13]:
stats.f.sf(32.03, dfn=3, dfd=16)


np.float64(5.376438762233098e-07)

In [1]:
# stats.f_oneway(group1, group2, group3, ...) does the same thing

# chi square analysis

In [27]:
# for confidence interval of 95%
stats.chi2.isf(0.05, df=3)


np.float64(7.814727903251178)

In [ ]:
from scipy.stats import chisquare

At an emporium, the manager is interested in knowing age group which visits the mall
during the day. He defines categories as - children, teenagers, adults and senior citizens.
He plans to have his inventory of goods accordingly. He claims that out of all the people
who visited 5% are children, 38% are teenagers, 2% are senior citizens are remaining are
adults.
From a sample of 180 people it was seen that 25 were children, 50 were teenagers, 90
were adults and 15 were senior citizens
Test the manager’s claim at 95% confidence level.


In [11]:
obs = [25, 50, 90, 15]
expected = [9, 68.4, 99, 3.6]

chisquare(obs, expected)

Power_divergenceResult(statistic=np.float64(70.31233386496544), pvalue=np.float64(3.659118590746897e-15))

Power_divergenceResult(statistic=np.float64(70.31233386496544), pvalue=np.float64(3.659118590746897e-15))
- statistic=np.float64(70.31233386496544) => distance
- pvalue=np.float64(3.659118590746897e-15) => probality that my assertion is true

obs = [25, 35, 15, 10]
exp = [50,15,10,10]

In [28]:
obs = [25, 35, 15, 10]
exp = [50,15,10,10]

distance = 0
for i in range(len(obs)):
    distance += ((obs[i]**2) / exp[i])
distance -= sum(obs)

distance = 0
for i, j in zip(obs, exp):
    distance += ((i**2) / j)
distance -= sum(obs)

print('digrees of freedom of the resultant chi square distribution = ', len(obs) - 1)
print('distance',distance)
print('cutoff point' , stats.chi2.isf(0.05, df=len(obs) - 1))

# since we want to reject on right hand side my pvalue is the the right side
print('pvalue', stats.chi2.sf(distance, df=len(obs) - 1))

chisquare(obs, exp)

digrees of freedom of the resultant chi square distribution =  3
distance 41.66666666666667
cutoff point 7.814727903251178
pvalue 4.721762209176316e-09


Power_divergenceResult(statistic=np.float64(41.66666666666667), pvalue=np.float64(4.721762209176316e-09))

if the pvalue is lesser than 5% then reject

or stats.chi.isf(0.05, df=3) > distance

note: statistic=np.float64(41.66666666666667) refers to distance

### aems ses 6 ppt
A study was conducted to test the effect of the malaria parasite - plasmodium falciparum -
on heterozygous and homozygous humans. The vaccine was given to a cohort of 252
humans. Test whether the heterozygous humans are better protected than homozygous.

In [ ]:
from scipy.stats import chi2_contingency

# H0 always assumes independence between the two variables
# H1 assumes dependence between the two variables

In [ ]:
obs = [[93, 51], [68, 40]]
chi2_contingency(obs, correction=False)
# pvalue > 0.05 accept H0 => they are independent

Chi2ContingencyResult(statistic=np.float64(0.07023411371237459), pvalue=np.float64(0.790996215494177), dof=1, expected_freq=array([[92., 52.],
       [69., 39.]]))

In [26]:
stats.chi2.sf(0.07023411371237459, df=1)

np.float64(0.790996215494177)

A psychologist wants study whether the happiness quotient of children in the house is
related to the family income. He collects data of 1300 children is there enough evidence to
claim that they are related.


In [22]:
obs = [[245, 354, 243], [98, 220, 140]]

chi2_contingency(obs, correction=False)

Chi2ContingencyResult(statistic=np.float64(9.372003254267797), pvalue=np.float64(0.009223491484580416), dof=2, expected_freq=array([[222.15846154, 371.77538462, 248.06615385],
       [120.84153846, 202.22461538, 134.93384615]]))

p<5% reject H0 => they are dependent

# Therefore money buys happiness